### PR TITLE
Feature/vdp 27 adding bbs to jsonld suites

### PR DESCRIPTION
**Pull request by maikel-maas**

### CODE REVIEW

File: packages/bls-kms-local/plugin.schema.json [hunk 1]
```diff
@@ -0,0 +1,473 @@
+{
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/plugin.schema.json:1]: i believe this file should be omitted

File: packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import {IKey} from "@veramo/core";
+import {BlsKeyManagementSystem} from "../agent/BlsKeyManagementSystem";
+
+describe('@sphereon/ssi-sdk-bls-kms-local', () => {
+    let bls: { publicKey: Uint8Array, secretKey: Uint8Array };
+    let kms: BlsKeyManager;
+
+    beforeAll(async () => {
+        bls = await generateBls12381G2KeyPair();
+        kms = new BlsKeyManager({ store: new MemoryKeyStore(),
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts:13]: why is this kms defined here? for rest you define it in the restAgent.tests and for local you define this in the agent.yaml

File: packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    })
+
+    it("should delete a bls key", async() => {
+        await expect(kms.keyManagerDelete({ kid: Buffer.from(bls.publicKey).toString('hex') })).resolves.toBeTruthy();
+    })
+
+    afterAll(async() => {
+        await new Promise((resolve) => setTimeout((v: void) => resolve(v), 500))
+    })
+})
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts:81]: newline

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+  private readonly privateKeyStore: AbstractPrivateKeyStore;
+
+  constructor(keyStore: AbstractPrivateKeyStore) {
+    super(keyStore);
+    this.privateKeyStore = keyStore;
+  }
+
+  async importKey(args: Exclude<MinimalImportableKey, 'kms'>): Promise<ManagedKeyInfo> {
+    if (args.type === <TKeyType>'Bls12381G2') {
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:20]: can we make this an early return? following these huge brackets is not nice :)

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      debug('imported key', managedKey.type, managedKey.publicKeyHex)
+      return managedKey
+    }
+    return super.importKey(args) as Promise<ManagedKeyInfo>
+  }
+
+  async createKey({ type }: { type: TKeyType }): Promise<ManagedKeyInfo> {
+    let key: ManagedKeyInfo
+    switch (type) {
+      case <TKeyType>'Bls12381G2': {
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:35]: would be nicer to make an enum for 'Bls12381G2'

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      default:
+        key = await super.createKey({ type })
+    }
+
+    debug('Created key', type, key.publicKeyHex)
+
+    return key
+  }
+
+  async sign({ keyRef, algorithm, data }: { keyRef: Pick<IKey, 'kid'>; algorithm?: string; data: Uint8Array | Uint8Array[] }): Promise<string> {
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:54]: would be nicer to define an interface for the return and use that here instead of the whole object

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        messages: Array.isArray(data)? data: [data] ,
+      }
+      return Buffer.from(
+        await blsSign(keyPair)
+      ).toString("hex");
+    }
+    throw Error(`not_supported: Cannot sign using key of type ${privateKey.type}`)
+  }
+
+  private asBlsManagedKeyInfo(args: { alias?: string, type: TKeyType, privateKeyHex: string, publicKeyHex: string }): ManagedKeyInfo {
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:79]: same comments here, please look at how veramo structures it's plugins and do the same. that way it is alot easier to understand other people's plugins

File: packages/bls-kms-local/src/types/IBlsKeyManager.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+export interface IKeyManagerVerifyArgs {
+    kms: string,
+    publicKey: Uint8Array,
+    messages: Uint8Array[],
+    signature: Uint8Array
+}
+
+export type BBSKey = Partial<IKey>
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/types/IBlsKeyManager.ts:79]: newline

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+export class SphereonBbsBlsSignature2020 extends SphereonLdSignature {
+  constructor() {
+    super()
+    // Ensure it is loaded
+    suiteContext2020?.constants
+  }
+
+  getSupportedVerificationType(): string {
+    return 'Bls12381G2Key2020'
```
Comment (BtencateSphereon) [inline packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:18]: create a enum for this

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  getSupportedVeramoKeyType(): TKeyType | 'Bls12381G2' {
+    return 'Bls12381G2'
```
Comment (BtencateSphereon) [inline packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:22]: create a enum for this

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      signer: () => signer,
+      type: this.getSupportedVerificationType(),
+    }
+    /*const key1 = new Bls12381G2KeyPair({
+      publicKeyBase58: hexToMultibase("", MultibaseFormat.BASE58).value
+    });
+    const bbsSignature = new BbsBlsSignature2020({
+      key1
+    })*/
+    let key2: Bls12381G2KeyPair = undefined as unknown as Bls12381G2KeyPair
```
Comment (BtencateSphereon) [inline packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:65]: i do not think the code here is complete

File: packages/bls-kms-local/src/agent/BlsKeyManager.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            throw Error(`invalid_argument: This agent has no registered KeyManagementSystem with name='${name}'`)
+        }
+        return kms
+    }
+
+    async keyManagerSign(args: Omit<IKeyManagerSignArgs, 'data'> & { data: string | string[]}): Promise<string> {
+        const keyInfo: IKey = await this.localStore.get({ kid: args.keyRef }) as IKey
+        const kms = this.getLocalKms(keyInfo.kms)
+        let input: any;
+        if (Array.isArray(args.data)) {
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/src/agent/BlsKeyManager.ts:30]: ternary would be nicer

File: packages/bls-kms-local/src/types/IBlsKeyManager.ts (hunk 8 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/types/IBlsKeyManager.ts:79]: Fixed!

File: packages/bls-kms-local/package.json [hunk 13]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "@sphereon/ssi-sdk-bls-key-manager",
+  "description": "Sphereon KMS implementation that provides BLS crypto",
+  "version": "0.5.0",
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/bls-kms-local/package.json:4]: Fixed

File: packages/bls-kms-local/plugin.schema.json (hunk 1 above)
Comment (maikel-maas) [reply packages/bls-kms-local/plugin.schema.json:1]: Removed

File: packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts (hunk 3 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts:81]: added

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 5 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:35]: TKeyType is an union enum and already contains the Bls12381G2 key type

Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:35]: Fixed anyway

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 6 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:54]: I'm using the exact interface that Veramo does. It probably makes more sense to add this changes to @veramo/key-manager and open a PR for them.

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 7 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:79]: Fixed!

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 9 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:18]: fixed

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 10 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:22]: TKeyType is an union enum and already contains Bls12381G2. Fixed anyway

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+    const options = {
+      id: id,
+      controller: controller,
+      privateKeyBase58: key.privateKeyHex as string,
+      publicKeyBase58: key.publicKeyHex,
+      signer: () => signer,
+      type: this.getSupportedVerificationType(),
+    }
+    /*const key1 = new Bls12381G2KeyPair({
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:59]: Removed

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 11 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:65]: Fixed

File: packages/bls-kms-local/src/agent/BlsKeyManager.ts (hunk 12 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManager.ts:30]: Fixed!

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 5 above)
Comment (BtencateSphereon) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts]: the point is to have 1 place where the text is placed so that you do not get miss types

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 6 above)
Comment (BtencateSphereon) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:59]: i am talking about the return type, refactor that to an interface and use that here

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 10 above)
Comment (BtencateSphereon) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts]: the point is to have 1 place where the text is placed so that you do not get miss types

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 5 above)
Comment (nklomp) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts]: Nah, that single point is actually this interface. Please stay as close to upstream as possible

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 10 above)
Comment (nklomp) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts]: Nah, that single point is actually this interface. Please stay as close to upstream as possible

File: packages/bls-kms-local/package.json [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    }
+  },
+  "scripts": {
+    "build": "tsc --build",
+    "generate-plugin-schema": "yarn veramo dev generate-plugin-schema"
+  },
+  "dependencies": {
+    "@mattrglobal/bbs-signatures": "^1.0.0",
+    "@mattrglobal/node-bbs-signatures": "^0.13.0",
+    "@veramo/core": "3.1.2-next.84"
```
Comment (nklomp) [inline packages/bls-kms-local/package.json:20]: Do we really need this version, because this is likely going to mess with other plugins

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 6 above)
Comment (nklomp) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:59]: Mwah, if we want to bring this upstream, stay in line with the way upstream does it right now. Changing things like this would need to require seperate PRs

File: packages/bls-kms-local/src/agent/BlsKeyManager.ts (hunk 12 above)
Comment (BtencateSphereon) [reply packages/bls-kms-local/src/agent/BlsKeyManager.ts]: haven't looked at the changes yet. But do not use any. If there are multiple types possible, state those multiple types. if you really have no clue what it could be, use unknown

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  async sign({ keyRef, algorithm, data }: { keyRef: Pick<IKey, 'kid'>; algorithm?: string; data: Uint8Array | Uint8Array[] }): Promise<string> {
+    let privateKey: ManagedPrivateKey
+    try {
+      privateKey = await this.privateKeyStore.get({ alias: keyRef.kid })
+    } catch (e) {
+      throw new Error(`key_not_found: No key entry found for kid=${keyRef.kid}`)
+    }
+
+    if (privateKey.type !== KeyType.Bls12381G2 && !Array.isArray(data)) {
+      return await super.sign({ keyRef, algorithm, data })
```
Comment (nklomp) [inline packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:68]: What if data is an array and type is not Bls?

File: packages/bls-kms-local/package.json [hunk 17]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "@sphereon/ssi-sdk-bls-key-manager",
+  "description": "Sphereon KMS implementation that provides BLS crypto",
```
Comment (BtencateSphereon) [inline packages/bls-kms-local/package.json:3]: Not sure this description really reflects what this plugin is about? it is about signing?

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+export class BlsKeyManagementSystem extends KeyManagementSystem {
+  private readonly privateKeyStore: AbstractPrivateKeyStore
+
+  constructor(keyStore: AbstractPrivateKeyStore) {
+    super(keyStore)
+    this.privateKeyStore = keyStore
+  }
+
+  async importKey(args: Exclude<MinimalImportableKey, 'kms'>): Promise<ManagedKeyInfo> {
+    if (args.type === KeyType.Bls12381G2) {
```
Comment (nklomp) [inline packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:20]: Change around the if / super flow using an early return to make it more clear that by default we call the super method

File: packages/did-auth-siop-op-authenticator/package.json [hunk 19]
```diff
@@ -16,16 +16,16 @@
     "@sphereon/did-auth-siop": "0.2.10",
     "@sphereon/pex": "^1.0.2",
     "@sphereon/ssi-sdk-core": "^0.5.0",
-    "@veramo/core": "3.1.1-next.26",
-    "@veramo/utils": "3.1.1-next.26",
+    "@veramo/core": "3.1.2-next.84",
```
Comment (nklomp) [inline packages/did-auth-siop-op-authenticator/package.json:19]: Is this really necessary. We are upping all our packages and I explicitly stopped at 26, because we needed to do some work to get past that. I doubt that all of a sudden resolved itself

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+
+    const options = {
+      id: id,
+      controller: controller,
+      privateKeyBase58: hexToMultibase(key.privateKeyHex, MultibaseFormat.BASE58).value.substring(1),
+      publicKeyBase58: hexToMultibase(key.publicKeyHex, MultibaseFormat.BASE58).value.substring(1),
+      type: this.getSupportedVerificationType(),
+    }
+
+    let key2: Bls12381G2KeyPair = new Bls12381G2KeyPair(options)
```
Comment (nklomp) [inline packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:49]: Please give this a more informative name

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+  getSuiteForSigning(key: IKey, issuerDid: string, verificationMethodId: string, context: IAgentContext<RequiredAgentMethods>): any {
+    const controller = issuerDid
+
+    const id = verificationMethodId
+
+    if (!key.privateKeyHex) {
+      throw new Error('Private key must be defined')
+    }
+
+    const options = {
```
Comment (nklomp) [inline packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:41]: Please give this a more informative name. The name options typically is an external method argument which makes this pretty confusing when you haven't started reading the code at the top

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+      privateKeyBase58: hexToMultibase(key.privateKeyHex, MultibaseFormat.BASE58).value.substring(1),
+      publicKeyBase58: hexToMultibase(key.publicKeyHex, MultibaseFormat.BASE58).value.substring(1),
+      type: this.getSupportedVerificationType(),
+    }
+
+    let key2: Bls12381G2KeyPair = new Bls12381G2KeyPair(options)
+
+    type KeyPairSignerOptions = { data: Uint8Array | Uint8Array[] }
+
+    const options2 = {
```
Comment (nklomp) [inline packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:53]: As soon as you are naming your variables 1, 2 or 3 something is up with your naming

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+    return new MattrBbsBlsSignature2020(options2)
+  }
+
+  private addMethodToBbsBlsSignature2020() {
+    function _includesContext(args: { document: any; contextUrl: string }) {
+      const context = args.document['@context']
+      return context === args.contextUrl || (Array.isArray(context) && context.includes(args.contextUrl))
+    }
+    //@ts-ignore
+    MattrBbsBlsSignature2020.prototype['ensureSuiteContext'] = function (args: { document: any; addSuiteContext: boolean }): void {
```
Comment (nklomp) [inline packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:79]: Please add comments when performing a hack

File: packages/vc-handler-ld-local/src/types/types.ts [hunk 24]
```diff
@@ -220,6 +225,11 @@ export interface IVerifyPresentationLDArgs {
    */
   presentationPurpose?: typeof ProofPurpose
 
+  /**
+   * Optional. The key handle ({@link IKey#kid}) from the internal database.
+   */
+  keyRef?: string
```
Comment (nklomp) [inline packages/vc-handler-ld-local/src/types/types.ts:231]: Why is this needed to verify a presentation?

File: packages/vc-handler-ld-local/src/types/types.ts [hunk 25]
```diff
@@ -175,6 +175,11 @@ export interface IVerifyCredentialLDArgs {
    */
   purpose?: typeof ProofPurpose
 
+  /**
+   * Optional. The key handle ({@link IKey#kid}) from the internal database.
+   */
```
Comment (nklomp) [inline packages/vc-handler-ld-local/src/types/types.ts:180]: Why do you need this to verify?

**Review by nklomp — CHANGES_REQUESTED** [review]

File: packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts (hunk 2 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts]: There is no agent at this point. Please check @veramo/kms-local/src/__tests__/kms-local.test.ts or @veramo/kms-local/src/__tests__/kms-local.test.ts

File: packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts [hunk 26]
```diff
@@ -0,0 +1,81 @@
+import {generateBls12381G2KeyPair} from '@mattrglobal/node-bbs-signatures';
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/bls-kms-local/src/__tests__/bls-key-manager.test.ts]: There is no agent at this point. Please check @veramo/kms-local/src/tests/kms-local.test.ts or @veramo/kms-local/src/tests/kms-local.test.ts

File: packages/vc-handler-ld-local/src/types/types.ts (hunk 24 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/types/types.ts]: Removed

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 23 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts:83]: Fixed!

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 22 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts]: Fixed!

File: packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts (hunk 21 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/src/suites/BbsBlsSignature2020.ts]: Fixed!

File: packages/did-auth-siop-op-authenticator/package.json (hunk 19 above)
Comment (maikel-maas) [reply packages/did-auth-siop-op-authenticator/package.json:19]: Actually Veramo adds partial support to bls keys from version 3.1.2-next.1 on. It would be nice to use the existing implementation.

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 18 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts]: Fixed!

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 16 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:70]: I will keep it as a single element to be compliant with the existing implementation.

File: packages/bls-kms-local/package.json (hunk 15 above)
Comment (maikel-maas) [reply packages/bls-kms-local/package.json:20]: Actually Veramo adds partial support to bls keys from version 3.1.2-next.1 on. It would be nice to use the existing implementation.

File: packages/bls-kms-local/package.json (hunk 17 above)
Comment (maikel-maas) [reply packages/bls-kms-local/package.json]: Changed the description, but we need to support bls keys not only in the key store but in did:key provider and resolver as well. Veramo has a partial implementation of it nowadays, so I'm not sure if I will create a PR for them or just wait until their changes are pushed to the upstream.

File: packages/bls-kms-local/src/agent/BlsKeyManager.ts (hunk 12 above)
Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManager.ts]: Just a dirty hack to be able to insert an array without changing the original interface. Removed the array, so I think it will work fine

File: packages/bls-kms-local/package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (nklomp) [reply packages/bls-kms-local/package.json]: Actually, since this is a mono repo using lerna, it should be in line with the base version, meaning 0.5.0. Lerna handles the versions. If this version deviates from the rest of the packages we get in trouble.

File: packages/bls-kms-local/package.json (hunk 15 above)
Comment (nklomp) [reply packages/bls-kms-local/package.json:20]: Seems like everything still works with the fixed integration test, so yes we can update

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 16 above)
Comment (nklomp) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:70]: Okay but then the change from `messages: Array.isArray(data) ? data : [data],` to `messages: [data],` should be reverted

Comment (nklomp) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:70]: Yeah then you can cast it right?  The whole point is that current change will fail if you submit an array. The code is too naive to expect data not to be an array already

Comment (maikel-maas) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:70]: There is no cast anymore and no interface changes either. Veramo does not accept Uint8Array[], but Uint8Array object and I'm adding the Uint8Arrat object to an array when passing it to the signer function. I added validation, just in case.

File: packages/bls-did-provider-key/src/BlsKeyDidResolver.ts [hunk 27]
```diff
@@ -0,0 +1,24 @@
+import { resolve } from '@transmute/did-key.js'
```
Comment (nklomp) [inline packages/bls-did-provider-key/src/BlsKeyDidResolver.ts:1]: What is the need or this whole module whilst Veramo already provides https://veramo.io/docs/api/did-provider-key.getdidkeyresolver/ ? Which is also described in our main readme: https://github.com/Sphereon-Opensource/ssi-sdk#did-resolution

File: packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts (hunk 16 above)
Comment (nklomp) [reply packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts:70]: Okay

**Review by nklomp — CHANGES_REQUESTED** [review]
Some remarks on latest changes. I fail to see why we need a new module for did key, whilst veramo already provides a method/function for it. Next to that a package for which is essentially a bit of confiiguration and dependency work is not really needed

File: packages/bls-did-provider-key/src/BlsKeyDidResolver.ts (hunk 27 above)
Comment (maikel-maas) [reply packages/bls-did-provider-key/src/BlsKeyDidResolver.ts]: Veramo's implementation does not resolve dids created with a bls keys. I had to add '@transmute/did-key-bls12381' and 'did:key:zUC7': resolveBls12381G2 to the startWithMap

Comment (nklomp) [reply packages/bls-did-provider-key/src/BlsKeyDidResolver.ts]: Okay. That would probably make a good PR upstream in the future.

**Review by nklomp — APPROVED** [review]

File: packages/vc-handler-ld-local/package.json [hunk 28]
```diff
@@ -21,17 +20,21 @@
     "@digitalcredentials/ed25519-verification-key-2020": "^3.2.2",
     "@digitalcredentials/jsonld": "^5.2.1",
     "@digitalcredentials/jsonld-signatures": "^9.3.1",
+    "jsonld-signatures": "5.0.1",
```
Comment (nklomp) [inline packages/vc-handler-ld-local/package.json:23]: Hzzz, can't we make it work with the above @Digtalcredentials version? This import will for sure fail with React Native. That is the whole reason we use the digital credentials forks

**Review by nklomp — CHANGES_REQUESTED** [review]
Overall looking good!

1 important question on the usage of jsonld. Rest is minor stuff

File: packages/vc-handler-ld-local/package.json (hunk 28 above)
Comment (maikel-maas) [reply packages/vc-handler-ld-local/package.json:23]: It will fail with React Native for sure: https://github.com/mattrglobal/jsonld-signatures-bbs/issues/177. There are major differences between this version and @digitalcredentials version. I tried to make it work and I got a VC with a non compliant proof and could not sign a VP at all. By the way, there is no version @digitalcredentials/jsonld-signature 5.0.1

File: packages/bls-key-manager/README.md [hunk 29]
```diff
@@ -0,0 +1,10 @@
+<!--suppress HtmlDeprecatedAttribute -->
+<h1 align="center">
+  <br>
+  <a href="https://www.sphereon.com"><img src="https://sphereon.com/content/themes/sphereon/assets/img/logo.svg" alt="Sphereon" width="400"></a>
+  <br>BLS KMS LOCAL
+  <br>
+</h1>
+
+This plugin combines multiple Key management systems and exposes a unified API to the DID manager
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/bls-key-manager/README.md:9]: Fixed
